packages/itwin/grouping-mapping-widget/src/widget/components/QueryBuilder.ts

```diff
@@ -157,333 +201,260 @@ export class QueryBuilder {
     return false;
   }
 
-  private addRelatedProperty(
+  private addRelatedToQuery(
     unionIndex: number,
     propertyField: PropertiesField,
     propertyName: string,
     propertyValue: Primitives.Value,
-    isAspect: boolean,
   ) {
-    const addProperty = this.addRelatedPropertyToQuery(unionIndex, isAspect);
-
     const paths = [...(propertyField.parent?.pathToPrimaryClass ?? [])];
     paths.reverse().forEach((path) => {
       const sourceClassName = path.sourceClassInfo?.name.replace(":", ".");
       const targetClassName = path.targetClassInfo?.name.replace(":", ".");
       const relClassName = path.relationshipInfo?.name.replace(":", ".");
 
-      const relPropertyName = path.isForwardRelationship
+      const relClassProperty = path.isForwardRelationship
         ? `SourceECInstanceId`
         : `TargetECInstanceId`;
 
       const relPropertyValue = path.isForwardRelationship
-        ? `${relClassName}.TargetECInstanceId`
-        : `${relClassName}.SourceECInstanceId`;
+        ? `TargetECInstanceId`
+        : `SourceECInstanceId`;
 
-      addProperty(
+      this.addClassToQuery(
+        unionIndex,
         targetClassName,
         `ECInstanceId`,
-        relPropertyValue,
-        false
+        relClassName,
+        relPropertyValue
       );
 
-      addProperty(
+      this.addClassToQuery(
+        unionIndex,
         relClassName,
-        relPropertyName,
-        `${sourceClassName}.ECInstanceId`,
-        false
+        relClassProperty,
+        sourceClassName,
+        `ECInstanceId`
       );
 
       if (path.sourceClassInfo?.name
         === propertyField.parent?.contentClassInfo.name) {
-        addProperty(
+        this.addPropertyToQuery(
+          unionIndex,
           sourceClassName,
           propertyName,
           propertyValue,
-          this.needsQuote(propertyField)
+          this.needsQuote(propertyField),
+          false,
+          false,
         );
       }
     });
   }
 
-  private addRelatedPropertyToQuery = (
+  private addClassToQuery(
     unionIndex: number,
-    isAspect: boolean,
-  ) => (
     className: string,
-    propertyName: string,
-    propertyValue: Primitives.Value,
-    needsQuote: boolean,
-  ) => this.addPropertyToQuery(unionIndex, className, propertyName, propertyValue, isAspect, needsQuote, false, true);
-
-  private addPropertyToQuery(
-    unionIndex: number,
-    className: string,
-    propertyName: string,
-    propertyValue: Primitives.Value,
-    isAspect: boolean,
-    needsQuote: boolean,
-    isCategory: boolean,
-    isRelational?: boolean,
+    classProperty: string,
+    joinClassName: string,
+    joinClassProperty: string,
   ) {
     if (this.query === undefined || this.query.unions.length === 0) {
```

Comment:
I guess the `this.query.unions.length === 0` check is meaningless here if you set it to an empty array afterwards.

Reply:
Removed here and in first `buildQueryString` check as it is the same situation there